tools/code-analyzer/src/git.ts

```diff
@@ -5,12 +5,135 @@ import { CliUx } from '@oclif/core';
+ * @return {string} - diff of the changfiles between the 2 hashes
+ */
+export const diffHashes = ( baseDir: string, hashA: string, hashB: string ) => {
+	const git = simpleGit( { baseDir } );
+	return git.diff( [ `${ hashA }..${ hashB }` ] );
+};
+
+/**
+ * Determines if a string is a commit hash or not.
+ *
+ * @param {string} ref - the ref to check
+ * @return {boolean} whether the ref is a commit hash or not.
+ */
+const refIsHash = ( ref: string ) => {
+	return /^[0-9a-f]{7,40}$/i.test( ref );
+};
+
+/**
+ * Get the commit hash for a ref (either branch or commit hash). If a validly
+ * formed hash is provided it is returned unmodified.
+ *
+ * @param {string} baseDir - the dir of the git repo to get the hash from.
+ * @param {string} ref     -    Either a commit hash or a branch name.
+ * @return {string} - the commit hash of the ref.
+ */
+export const getCommitHash = async ( baseDir: string, ref: string ) => {
+	const isHash = refIsHash( ref );
+
+	// If its not a hash we assume its a branch
+	if ( ! isHash ) {
+		return simpleGit( { baseDir } ).revparse( [ ref ] );
+	}
+
+	// Its a hash already
+	return ref;
+};
+
+/**
+ * generateDiff generates a diff for a given repo and 2 hashes or branch names.
+ *
+ * @param {string}   repoPath - the url or filepath of the repo to clone.
+ * @param {string}   hashA    - commit hash or branch name.
+ * @param {string}   hashB    - commit hash or branch name.
+ * @param {Function} onError  - the handler to call when an error occurs.
+ */
+export const generateDiff = async (
+	repoPath: string,
+	hashA: string,
+	hashB: string,
+	onError: ( error: string ) => void
+) => {
+	try {
+		const context = await cloneRepo( repoPath );
+		const tmpRepoPath = join( tmpdir(), context );
+
+		await simpleGit( { baseDir: tmpRepoPath } ).pull( [ '--all' ] );
+
+		// checking out any branches will automatically track remote branches.
+		await Promise.all(
+			[ hashA, hashB ]
+				.filter( ( hash ) => ! refIsHash( hash ) )
+				.map( ( branch ) => {
+					return simpleGit( { baseDir: tmpRepoPath } ).checkout( [
+						branch,
+					] );
+				} )
+		);
+
+		// turn both hashes into commit hashes if they are not already.
+		const commitHashA = await getCommitHash( tmpRepoPath, hashA );
+		const commitHashB = await getCommitHash( tmpRepoPath, hashB );
+
+		const isRepo = await simpleGit( {
+			baseDir: tmpRepoPath,
+		} ).checkIsRepo();
+
+		if ( ! isRepo ) {
+			throw new Error( 'Not a git repository' );
+		}
+
+		const diff = await diffHashes( tmpRepoPath, commitHashA, commitHashB );
+
+		// time to clean up
+		rmSync( tmpRepoPath, { force: true, recursive: true } );
+
+		return diff;
+	} catch ( e ) {
+		if ( e instanceof Error ) {
+			onError(
+				`Unable to create diff. Check that git repo, base hash, and compare hash all exist.\n Error: ${ e.message }`
+			);
+		} else {
+			onError(
+				'Unable to create diff. Check that git repo, base hash, and compare hash all exist.'
+			);
+		}
+
+		return '';
+	}
+};
 
```

Comment:
I've left the existing git code in place, but migrating the db schema scan in a follow up PR should allow us to remove most of it.